File: frontend/lib/features/driver/services/driver_service.dart
import '../domain/repositories/driver_repository.dart';
import '../domain/enums/driver_availability.dart';
import '../domain/enums/driver_status.dart';
import '../domain/entities/driver.dart';

class DriverService {
  final DriverRepository _repository;

  DriverService(this._repository);

  Future<Driver> getDriver(String driverId) async {
    return _repository.getDriverById(driverId);
  }

  Future<void> goAvailable(String driverId) async {
    final driver = await getDriver(driverId);
    if (driver.currentStatus != DriverStatus.idle && driver.currentStatus != DriverStatus.completed) {
      throw Exception('Cannot go available while currently assigned to an emergency');
    }
    await _repository.updateAvailability(driverId, DriverAvailability.available);
  }

  Future<void> updateEmergencyStatus(String driverId, DriverStatus newStatus) async {
    final driver = await getDriver(driverId);
    if (driver.availabilityStatus == DriverAvailability.offline) {
      throw Exception('Driver is currently offline');
    }
    await _repository.updateStatus(driverId, newStatus);
  }
}

File: backend/app/services/driver_service.py
from typing import List, Optional
from datetime import datetime
from ..models.driver import DriverModel
from ..schemas.driver import DriverCreate, DriverAvailabilityEnum, DriverStatusEnum
from ..repositories.driver_repository import DriverRepository

class DriverService:
    def __init__(self, repository: DriverRepository):
        self.repository = repository

    async def create_driver(self, request_data: DriverCreate) -> DriverModel:
        now = datetime.utcnow()
        new_driver = DriverModel(
            driverId=request_data.driverId,
            fullName=request_data.fullName,
            phoneNumber=request_data.phoneNumber,
            licenseNumber=request_data.licenseNumber,
            ambulanceId=request_data.ambulanceId,
            availabilityStatus=DriverAvailabilityEnum.OFFLINE.value,
            currentStatus=DriverStatusEnum.IDLE.value,
            createdAt=now,
            updatedAt=now
        )
        return await self.repository.create(new_driver)

    async def get_driver(self, driver_id: str) -> Optional[DriverModel]:
        return await self.repository.get_by_driver_id(driver_id)

    async def get_available_drivers(self) -> List[DriverModel]:
        return await self.repository.get_available_drivers()

    async def update_availability(self, driver_id: str, availability: str) -> Optional[DriverModel]:
        driver = await self.get_driver(driver_id)
        if not driver:
            raise ValueError("Driver not found")

        if driver.currentStatus not in [DriverStatusEnum.IDLE.value, DriverStatusEnum.COMPLETED.value] and availability == DriverAvailabilityEnum.AVAILABLE.value:
            raise ValueError("Cannot set to available while on an active emergency")

        update_data = {
            "availabilityStatus": availability,
            "updatedAt": datetime.utcnow()
        }
        return await self.repository.update(driver.id, update_data)

    async def update_status(self, driver_id: str, new_status: str) -> Optional[DriverModel]:
        driver = await self.get_driver(driver_id)
        if not driver:
            raise ValueError("Driver not found")

        if driver.availabilityStatus == DriverAvailabilityEnum.OFFLINE.value:
            raise ValueError("Offline drivers cannot have status updates")

        return await self.repository.update_status(driver_id, new_status)
